### DISCUSSION TITLE
Updating Data Using useFetch

### DISCUSSION THREAD

**Discussion by harblaith7** [nuxt/nuxt#18082] [Questions] [answered]
I am having issues updating my data when using useFetch. This is what I am trying to do:

```javascript
 async setup() {
    const city = ref("");
    const {
      data: weather,
      refresh,
    } = await useFetch(
      `https://api.openweathermap.org/data/2.5/weather?q=${
        city.value ? city.value : "Toronto"
      }&units=metric&appid=${apiKey}`
    );

  return {city, weather, refresh}
 }
```

```vue
        <input
          class="w-1/2 h-10"
          type="text"
          placeholder="Find a city..."
          v-model="city"
        />
        <button class="bg-sky-400 w-20 text-white h-10" @click="refresh">
          Search
        </button>
```

The initial fetch works, but when I type a new city and click search, it doesn't call the API with the correct url and update the data

**Comment by Diizzayy** [comment]
@harblaith7 
>  it doesn't call the API with the correct url and update the data

Is the subsequent request made at all? if so, is it that the city value isn't being updated?

**Reply by harblaith7** [reply]
No the call is being made, but with Toronto and not the new city. I am trying to find a way to use `useFetch` multiple to get called when triggered and update the data

**Reply by harblaith7** [reply]
I think refresh just recalls the first url and doesn't update

**Reply by Diizzayy** [reply]
Understood

**Comment by Diizzayy** [comment, marked as answer]
@harblaith7 

`useFetch`, it automatically generates a key based on URL given hence the URL cannot then be changed dynamically.

I've updated your code below to use `<script setup>` and `useAsyncData`. It should now work as you'd expect.

```
<script lang="ts" setup>
const apiKey = ref("");
const city = ref("");
const { data: weather, refresh } = await useAsyncData("weather", () =>
  $fetch(
    `https://api.openweathermap.org/data/2.5/weather?q=${
      city.value ? city.value : "Toronto"
    }&units=metric&appid=${apiKey.value}`
  )
);
</script>
```

**Reply by harblaith7** [reply]
That works! I understand why we have to change it to useAsyncData, but why did we have to change it to `<script setup>`? I tried running your code inside the setup() method, and it didn't work. Any ideas why?

```javascript
<script>
import { defineComponent, ref } from "vue";

export default defineComponent({
  async setup() {
    const city = ref("");
    const { data: weather, refresh } = await useAsyncData("weather", () =>
      $fetch(
        `https://api.openweathermap.org/data/2.5/weather?q=${
          city.value ? city.value : "Toronto"
        }&units=metric&appid=${myApiKey}`
      )
    );
    return {
      weather,
      refresh,
    };
  },
});
</script>
```

**Reply by Diizzayy** [reply]
using `<script setup>` isn't required, it's just my personal preference and it just simplifies the code.

The only issue I see with your code is that you forgot to return city in the setup function.

Also when using Nuxt3, you don't need to manually `import { defineComponent, ref } from "vue"`, that entire line can be omitted and they'd be imported by Nuxt3 automatically.


Your updated code:
```
<script>
export default defineComponent({
  async setup() {
    const city = ref("");
    const { data: weather, refresh } = await useAsyncData("weather", () =>
      $fetch(
        `https://api.openweathermap.org/data/2.5/weather?q=${
          city.value ? city.value : "Toronto"
        }&units=metric&appid=${myApiKey}`
      )
    );
    return { city, weather, refresh };
  },
});
</script>
```

**Reply by harblaith7** [reply]
Yup that was it. Thanks a ton
